Feat/toggle admin status

IanFonzie opened the pull request:
This PR closes issue: [DMM-240, DMM-265, DMM-266, DMM-267, DMM-268, DMM-269, DMM-270, DMM-271]

Includes tests? [N]
Updated docs? [N]

Proposed changes:
- Modifies Affiliations update endpoint to accept multiple types of update requests
- Adds a new "updateAdminStatus" update request type
- Adds a checkbox to each member on the team team page that can only be used according to the user stories referenced in the issues above
- Display confirmation modal when granting admin rights

bdolor commented on src/front-end/typescript/lib/pages/user/profile/tab/organizations.tsx:
DMM-267 states that a "Vendor Admin should be able to delegate Admin functionality to my team members" which requires the ability to edit an organization, so I would expect this requires a link to edit organizations in the 'Affiliated Organizations' table, (function `affiliatedTableBodyRows`) similar to 'Owned Organizations', (function `ownedTableBodyRows`)

Would also require updating permissions to read/GET `api/affiliations?organizations` endpoint

![image](https://github.com/bcgov/digital_marketplace/assets/2048170/aebf7fb8-cce9-4b46-bd47-bb332c4774c0)

IanFonzie replied:
Fair enough, I was going to leave this part to a different PR to cut down on size, but I think the way you've framed it makes sense

bdolor commented on src/back-end/lib/permissions.ts:
would look at adding permissions to line 228 so that Vendor Admins can see membership lists
```
    (!!session && (await isUserOwnerOfOrg(connection, session.user, orgId) ||
      (await isUserAdminOfOrg(connection, session.user, orgId))))
  );
  ```

IanFonzie replied:
Made some changes that should simplify these permissions in [32da1c2](https://github.com/bcgov/digital_marketplace/pull/429/commits/32da1c21f23e922e5856f0d9d1614668478ed0a8) and [4c07f2c](https://github.com/bcgov/digital_marketplace/pull/429/commits/4c07f2ceae5421484633b4639ef0270286a359bd)

bdolor reviewed (changes requested):
currently, vendor admins can't see the team list due to permission constraints, nor do they have a way to navigate to the teams list to make modifications as per DMM-268. Once permissions are fixed they can remove their own Admin functionality which is likely undesirable.

IanFonzie commented on src/front-end/typescript/lib/pages/organization/edit/tab/team.tsx:
```diff
@@ -441,6 +537,34 @@ function membersTableBodyRows(
+                    );
+              }}
+              disabled={isLoading || memberIsOwner(m)}
+              checked={isAffiliationAdminStatusChecked(m)}
```
Used an onChange event instead, I think I was messing around with the onClick and forgot to change it back

IanFonzie commented:
@bdolor, Addressed your changes and also added in support for [DMM-240](https://bcdevex.atlassian.net/browse/DMM-240). It's not addressed in the user stories but [85c9a86](https://github.com/bcgov/digital_marketplace/pull/429/commits/85c9a86c244b50ffc09b0cdc1325c13532ac6160) makes org links clickable for admins from the dashboard. It might be a good idea for us to go over those changes together as a sanity check.

bdolor commented on src/front-end/typescript/lib/pages/user/profile/tab/organizations.tsx:
```diff
@@ -420,7 +421,16 @@ function affiliatedTableBodyRows(
         children: (
           <div>
-            <span>{affiliation.organization.legalName}</span>
+            {memberIsAdmin(affiliation) ? (
```
could consider more specific naming convention as you have done in the back-end to disambiguate from application Admins, with `memberIsOrgAdmin`

IanFonzie replied:
good idea

bdolor commented on src/shared/lib/resources/affiliation.ts:
```diff
@@ -95,3 +99,15 @@ export function memberIsOwner(
   return member.membershipType === MembershipType.Owner;
 }
+
+export function memberIsAdmin(
```
maybe `memberIsOrgAdmin`?

IanFonzie commented on src/front-end/typescript/lib/components/accept-org-admin-terms.tsx:
```diff
@@ -0,0 +1,85 @@
+};
+
+export const update: component_.base.Update<State, Msg> = ({ state, msg }) => {
+  switch (msg.tag) {
```
The state ended up being unnecessary, so I removed it
